exercises/01.compartmentalisation/js.cc

```diff
@@ -0,0 +1,568 @@
+
+/**
+ * Code related to storing a secret value.  The attacker should try to leak
+ * this.
+ */
+namespace
+{
+	/// The secret value.
+	int32_t secret;
+
+	/**
+	 * Set the secret the first time.
+	 *
+	 * This blocks until there is a character available on the UART.  This is
+	 * *not* a good way of getting entropy, but it's the only one that we have
+	 * in the simulator.
+	 */
+	void set_secret()
+	{
+		auto uart = MMIO_CAPABILITY(Uart, uart);
+		while (!uart->can_read()) {}
+		secret = static_cast<int32_t>(rdcycle64());
+		Debug::log("Secret stored at {}", &secret);
+	}
+
+	/**
+	 * Report the current value of the secret and pick a new one.
+	 */
+	void check_secret(int32_t guess)
+	{
+		Debug::log("Secret was {}, you guessed {}.", secret, guess);
+		if (guess == secret)
+		{
+			Debug::log("CONGRATULATIONS! You correctly leaked the secret!");
+		}
+		secret = static_cast<int32_t>(rdcycle64());
+	}
+} // namespace
+
+/**
+ * Code related to the JavaScript interpreter.
+ */
+namespace
+{
+	/**
+	 * Constants for functions exposed to JavaScript->C++ FFI
+	 *
+	 * The values here must match the ones used in cheri.js.
+	 */
+	enum Exports : mvm_HostFunctionID
+	{
+		Print = 1,
+		Move,
+		LoadCapability,
+		LoadInt,
+		Store,
+		GetAddress,
+		SetAddress,
+		GetBase,
+		GetLength,
+		GetPermissions,
+		CheckSecret,
+	};
+
+	/// Constant for the say_hello function exposed to C++->JavaScript FFI
+	static constexpr mvm_VMExportID ExportRun = 1234;
+
+	/**
+	 * Type used for the set of capabilities that JavaScript has complete
+	 * control over.
+	 */
+	using VMRegisterState = std::array<void *, 8>;
+
+	/**
+	 * Address of the `VMRegisterState` on the stack.  We store this on the
+	 * stack so that it can hold local capabilities.
+	 */
+	ptraddr_t vmRegisterStateAddress;
+
+	/**
+	 * Re-derive the pointer to the on-stack register state.
+	 */
+	VMRegisterState &state()
+	{
+		register VMRegisterState *cspRegister asm("csp");
+		asm("" : "=C"(cspRegister));
+		Capability<VMRegisterState> rs{cspRegister};
+		rs.address() = vmRegisterStateAddress;
+		rs.bounds()  = sizeof(VMRegisterState);
+		return *rs;
+	}
+
+	/**
+	 * Write a capability into one of the VM's register set.
+	 */
+	void register_write(int regno, void *value)
+	{
+		if ((regno >= 0) && (regno < 8))
+		{
+			state()[regno] = value;
+		}
+	}
+
+	/**
+	 * Read a register from the VM's register set.  This reads one of the 8
+	 * that can be written or CSP (8), CGP (9), or PCC (10).
+	 */
+	void *register_read(int regno)
+	{
+		switch (regno)
+		{
+			default:
+				return nullptr;
+			case 0 ... 7:
+				return state()[regno];
+			case 8:
+			{
+				register void *cspRegister asm("csp");
+				asm("" : "=C"(cspRegister));
+				return cspRegister;
+			}
+			case 9:
+			{
+				register void *cgpRegister asm("cgp");
+				asm("" : "=C"(cgpRegister));
+				return cgpRegister;
+			}
+			case 10:
+			{
+				void *pcc;
+				asm("auipcc %0, 0\n" : "=C"(pcc));
+				return pcc;
+			}
+		}
+	}
+
+	/**
```

Comment:
Could all of this, through `call_export` be factored out into a microvium glue header or even pushed upstream?  It feels a little generic and/or out of place here.